doc/tutorials/advanced-ensemble.md

```diff
@@ -0,0 +1,603 @@
+httpd-v0
+root:
+"www"
+listen:
+("0.0.0.0:8080")
+handlers:
+(("/handler" . :demo/handler))
+enable-servlets:
+#t
+```
+
+And now let's start it and interact with it.
+```
+1$ gerbil httpd -G project/.gerbil server
+...
+
+2$ curl http://localhost:8080/
+2$ curl http://localhost:8080/servlets/hello.ss
+2$ curl http://localhost:8080/handler
+```
+
+### Using an httpd ensemble
+
+A single server is just that: one process, with one thread of control
+serving requests.  If we want to take advantage of multicore machines
+with process isolation, we can spawn more and multiplex on the socket
+using `SO_REUSEPORT` (the server does it by default).
+
+We can do this very easily by constructing an ensemble, and by doing
+that we take advantage of supervision and the all the available
+tooling from `gxensemble`.
+
+Here is how we can configure an httpd ensemble:
+- we configure the ensemble root directory
+- we configure the ensemble domain and worker domain
+- we configure the number of workers
+- and just pass the through the httpd config.
+
+```
+1$ pushd site
+1$ gerbil httpd -G project/.gerbil config --ensemble --ensemble-root root --ensemble-domain /test --worker-domain /test/www -n 2 --root www --enable-servlets --handlers '(("/handler" . :demo/handler))'
+
+1$ gerbil httpd -G project/.gerbil config --ensemble --print
+config:
+ensemble-v0
+roles:
+((httpd server-config:
+        (config:
+         ensemble-server-v0
+         application:
+         ((httpd config:
+                 httpd-v0
+                 root:
+                 "www"
+                 listen:
+                 ("0.0.0.0:8080")
+                 handlers:
+                 (("/handler" . :demo/handler))
+                 enable-servlets:
+                 #t))
+         env:
+         #f)
+        exe:
+        "gerbil"
+        prefix:
+        ("httpd" "server")
+        policy:
+        restart))
+preload:
+(workers: ((/test/www prefix: httpd role: httpd servers: 2)))
+domain:
+/test
+root:
+"root"
+```
+
+This configuration will preload and supervise, restarting as needed, 2 workers.
+The all we have to do is:
+```
+1$ gerbil httpd -G project/.gerbil ensemble
+...
+```
+
+So let's interact with our ensemble, and notice the multiplexing by
+the server identifier in the response:
+```
+2$ ps auxw | grep gerbil
+2$ curl http://localhost:8080/
+2$ curl http://localhost:8080/servlets/hello.ss
+2$ curl http://localhost:8080/handler
+
+```
+
+## Local Ensembles
+
+What about services composed of things other than httpds? Can't we run
+a general supervised ensemble that is open ended and can dynamically
+add or remove servers?
+
+The answer is a resounding yes; in fact `gerbil httpd ensemble` runs a
+generic supervisor and which just happens to be configured for httpd.
+In the folling we see how we can run an httpd ensemble with a standalone supervisor.
+
+First, let's configure the ensemble:
+- first we configure the httpd as usual -- notice the use of `-G env/local` so that
```

Comment:
`-G env/local` -> lowercase 'g'?

Reply:
no, we use that for a flat that says do global .gerbil